s3/pkg/service/object.go

```diff
@@ -256,9 +258,91 @@ func (s *s3Service) PutObject(ctx context.Context, in pb.S3_PutObjectStream) err
 	return nil
 }
 
-func (s *s3Service) GetObject(ctx context.Context, in *pb.GetObjectInput, out *pb.Object) error {
+func (s *s3Service) GetObjectMeta(ctx context.Context, in *pb.Object, out *pb.GetObjectMetaResult) error {
+	var err error
+	defer func() {
+		out.ErrorCode = GetErrCode(err)
+	} ()
+
+	object, err := s.MetaStorage.GetObject(ctx, in.BucketName, in.ObjectKey, true)
+	if err != nil {
+		log.Errorln("failed to get object info from meta storage. err:", err)
+		return err
+	}
+	out.Object = object.Object
+	return nil
+}
+
+func (s *s3Service) GetObject(ctx context.Context, req *pb.GetObjectInput, stream pb.S3_GetObjectStream) error {
 	log.Infoln("GetObject is called in s3 service.")
+	bucketName := req.Bucket
+	objectName := req.Key
+	offset := req.Offset
+	length := req.Length
+
+	var err error
+	getObjRes := &pb.GetObjectResponse{}
+	defer func ()  {
+		getObjRes.ErrorCode = GetErrCode(err)
+		stream.SendMsg(getObjRes)
+	} ()
+
+	object, err := s.MetaStorage.GetObject(ctx, bucketName, objectName, true)
+	if err != nil {
+		log.Errorln("failed to get object info from meta storage. err:", err)
+		return err
+	}
+
+	bucket, err := s.MetaStorage.GetBucket(ctx, bucketName, true)
+	if err != nil {
+		log.Errorln("failed to get bucket from meta storage. err:", err)
+		return err
+	}
+
+	// if this object has only one part
+	backend, err := getBackend(ctx, s.backendClient, bucket)
+	if err != nil {
+		log.Errorln("unable to get backend. err:", err)
+		return err
+	}
+	sd, err := driver.CreateStorageDriver(backend.Type, backend)
+	if err != nil {
+		log.Errorln("failed to create storage driver. err:", err)
+		return nil
+	}
+	log.Infof("get object offset %v, length %v", offset, length)
+	reader, err := sd.Get(ctx, object.Object, offset, offset + length - 1)
+	if err != nil {
+		log.Errorln("failed to get data. err:", err)
+		return err
+	}
+
+	eof := false
+	left := object.Size
+	buf := make([]byte, ChunkSize)
+	for !eof && left > 0 {
+		n, err := reader.Read(buf)
+		if err != nil && err != io.EOF {
+			log.Errorln("failed to read, err:", err)
+			break
```

Comment:
Do we need to return this error to the api service in some way?

Reply:
I dont sure.  there is a situation that api service have sent some data to end user, and return the error to api service, it do nothing except print it

Reply:
Then how to let the end user know there is something wrong when getting the object? Is it necessary?

Reply:
because http header status code in response could has been sent to end user,  we can not send error info to user, and the request is failed because of partial download

Reply:
The request will failed for partial download, and end user will receive a partial download error.  It is enough to print error info